packages/flutter_tools/lib/src/commands/daemon.dart

```diff
@@ -456,9 +457,15 @@ class AppDomain extends Domain {
     if (app == null)
       throw "app '$appId' not found";
 
-    return app._runInZone(this, () {
+    if (!fullRestart && inProgressHotReload != null)
```

Comment:
We probably want to disallow any concurrent reload/restart? So, `if (_inProgressReload != null) ...`

Reply:
Changed in b01335d. I wasn't totally sure if we wanted to affect full restarts too, but I guess it could suffer from similar issues so seems safest (as long as we don't ever go leaving this future around when it should've been cleared - I think hot reload has a timeout so hopefully this is all handled nicely).